### ISSUE TITLE
Segfault when using Distances

### ISSUE THREAD

**Issue by baggepinnen** [FluxML/Zygote.jl#385]
```julia
using Distances
using Flux
using Flux: gradient
Xp = randn(500)
X = randn(500)
lossobj = CosineDist()
lossobj(X,Xp)
loss = Xp -> evaluate(lossobj, X,Xp)
gradient(loss, Xp)
```
results in 
https://pastebin.com/ixVm1dSd
The error does not appear if `Xp,X` are scalars.

**Comment by baggepinnen** [comment]
I do think it is supposed to work, at least the [documentation says that custom structs should be no issue](https://fluxml.ai/Zygote.jl/latest/#Structs-and-Types-1). Why does there even have to be explicit gradients for Distances, other than for increased performance? Isn't Zygote supposed to figure the gradient out automatically? In any case, a segmentation fault is most definitely a bug. 

On a side, is having custom adjoints for every function in every ecosystem package really a sustainable strategy? If one is implemented every time a bug in Zygote is exposed, rather than fixing that bug, Zygote will always remain buggy?

**Comment by willtebbutt** [comment]
Don't get me wrong, I completely agree that it shouldn't be segfault-ing, clearly a bug. If it's mutating stuff internally though (and presumably it is) then we shouldn't expect Zygote to work.

As regards the segfault, what Julia version are you running on?

**Comment by baggepinnen** [comment]
I have tried on both julia v1.3-rc3 and v1.3-rc4
The segfault occurs for `Euclidean` as well as for `CosineDist`

**Comment by willtebbutt** [comment]
@MikeInnes any thoughts on the segfault? Something you've seen before?

**Comment by baggepinnen** [comment]
I now get a very similar error for a completely different code, the beginning of the stack trace looks like
```
Internal error: encountered unexpected error in runtime:
BoundsError(a=Array{Any, (2,)}[
  Core.Compiler.VarState(typ=Zygote.Pullback{Tuple{typeof(StaticArrays.unroll_tuple), Array{Base.Complex{Float64}, 1}, StaticArrays.Length{2}}, Any}, undef=false),
  Core.Compiler.VarState(typ=Array{Base.Complex{Float64}, 1}, undef=false)], i=(3,))
rec_backtrace at /buildworker/worker/package_linux64/build/src/stackwalk.c:94
record_backtrace at /buildworker/worker/package_linux64/build/src/task.c:224 [inlined]
jl_throw at /buildworker/worker/package_linux64/build/src/task.c:461
jl_bounds_error_ints at /buildworker/worker/package_linux64/build/src/rtutils.c:183
setindex! at ./essentials.jl:451 [inlined]
stupdate! at ./compiler/typelattice.jl:243
typeinf_local at ./compiler/abstractinterpretation.jl:1233
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1260
```

**Comment by CarloLucibello** [comment]
Differentiating Distances metrics works (via the `ZygoteDistancesExt` extension) and matches ForwardDiff; the segfault is gone. Closing as fixed.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
